### ISSUE TITLE
Feature Request: hs.brightness.watcher

### ISSUE THREAD

**Issue by TSFoster** [Hammerspoon/hammerspoon#853]
Hi,

I was thinking of scripting a switch-to-dark-mode-when-screen-brightness-is-low feature (like Tweetbot for iOS has) for my laptop, but there's not a watcher for `hs.brightness`.

I know I could set something up with `hs.timer.doEvery` or something, but I wondered if others thought `hs.brightness.watcher` might be useful and not problematic to implement?

**Comment by cmsj** [comment]
I'm fairly sure there's no system API for us to receive notifications when the screen brightness changes, so it would likely have to be something that polls with `hs.timer.doEvery`.

I'd love to be wrong though, if someone knows of maybe an NSDistributedNotificationCenter message that gets sent :)

**Comment by lowne** [comment]
Only tangentially related rant, but for future reference:

- `hs.redshift` has a full-blown invert mode, because
  1. the vast majority of Mac apps don't give a crap about the system-wide dark mode
  2. some have clunky ways to toggle a dark mode manually (Chrome extensions, Eclipse themes, iTerm themes, ...) but a hypothetical module to automate all that on a keystroke would be a pain to write and maintain (not to mention needing restarting apps etc.)
  3. since iOS 7 happened apps keep getting closer (by means of "new modern design!") to today's ideal colour scheme of pale white on white (a hilarious example is Evernote: people have been requesting a dark mode for ages, and what they got instead is vast expanses of white, inexorably growing with every new release, engulfing more and more of the UI)
- #694 polls the ambient light sensor (via `hs.timer`) to adjust the brightness on external screens, and can toggle inversion (via `hs.redshift`) when it's dark

TL;DR: I've "configured" my apps (takes literally 0 effort for 99% of them) to fully glow in Jony's endless, unapologetic whiteness - which is fine in bright daylight - but my retinas don't get destroyed at night/in the dark.

**Comment by TSFoster** [comment]
Thanks for getting back to me quickly, I figured there was probably a reason it didn't exist already :smiley:.

I ended up adding this to my Karabiner settings:

```xml
  <item>
    <name>Capture brightness changes</name>
    <appendix>Presses F18 in addition to brightness keys</appendix>
    <identifier>io.tsf.capture_brightness</identifier>
    <autogen>__KeyToKey__ ConsumerKeyCode::BRIGHTNESS_UP, ConsumerKeyCode::BRIGHTNESS_UP, KeyCode::F18</autogen>
    <autogen>__KeyToKey__ ConsumerKeyCode::BRIGHTNESS_DOWN, ConsumerKeyCode::BRIGHTNESS_DOWN, KeyCode::F18</autogen>
  </item>
```

so that a function key was triggered each time I changed brightness (I have the "Automatically adjust brightness" setting turned off, so manually adjusting is the only way the brightness changes). So now my dark mode script is just triggered by the F18 keypress:

```lua
hs.hotkey.bind({}, "f18", setDarkModeFromBrightness)
```
